File: tape/sdk/go/chaos/connectors.go
package chaos

import (
	"context"
	"math/rand/v2"
	"time"

	"github.com/vamsiramakrishnan/durable-agents/tape/sdk/go/connectors"
)

// ChaosConnector — wraps an inner Connector with declarative faults.
// Mirrors `tape.chaos.ChaosConnector` (Python) and `chaos.ChaosConnector`
// (TS). The three fault kinds it consumes are lose_ack, duplicate, delay.
type ChaosConnector struct {
	Inner  connectors.Connector
	Faults []Fault
	rng    *rand.Rand
}

// WrapConnector — sugar to build a ChaosConnector with a given RNG. Pass
// the Session's RNG so two replays with the same seed produce the same
// fault sequence.
func WrapConnector(inner connectors.Connector, faults []Fault, rng *rand.Rand) *ChaosConnector {
	if rng == nil {
		rng = rand.New(rand.NewPCG(rand.Uint64(), rand.Uint64()))
	}
	return &ChaosConnector{Inner: inner, Faults: faults, rng: rng}
}

// Name — forward to the inner connector so the registry routes correctly.
func (c *ChaosConnector) Name() string { return c.Inner.Name() }

func (c *ChaosConnector) fire(kind string) *Fault {
	for i := range c.Faults {
		f := c.Faults[i]
		if f.Action != kind {
			continue
		}
		if f.Probability >= 1.0 || c.rng.Float64() < f.Probability {
			return &c.Faults[i]
		}
	}
	return nil
}

func (c *ChaosConnector) Dispatch(ctx context.Context, e connectors.Effect) (connectors.DispatchResult, error) {
	if d := c.fire("delay"); d != nil && d.Ms > 0 {
		ms := float64(d.Ms)
		if d.Jitter > 0 {
			ms = ms * (1.0 + (c.rng.Float64()*2-1)*d.Jitter)
			if ms < 0 {
				ms = 0
			}
		}
		t := time.NewTimer(time.Duration(ms) * time.Millisecond)
		select {
		case <-ctx.Done():
			t.Stop()
			return connectors.DispatchResult{}, ctx.Err()
		case <-t.C:
		}
	}
	result, err := c.Inner.Dispatch(ctx, e)
	if err != nil {
		return result, err
	}
	if result.Outcome == connectors.DispatchConfirmed && c.fire("lose_ack") != nil {
		return connectors.DispatchResult{
			Outcome:    connectors.DispatchUnknown,
			Response:   result.Response,
			Error:      "tape.chaos: simulated lost ack",
			DispatchID: result.DispatchID,
		}, nil
	}
	return result, nil
}

func (c *ChaosConnector) Observe(ctx context.Context, e connectors.Effect) (connectors.ObservationResult, error) {
	result, err := c.Inner.Observe(ctx, e)
	if err != nil {
		return result, err
	}
	if result.Outcome == connectors.ObservationConfirmed && c.fire("duplicate") != nil {
		return connectors.ObservationResult{
			Outcome:  connectors.ObservationDuplicate,
			Response: result.Response,
			Count:    result.Count + 1,
		}, nil
	}
	return result, nil
}

func (c *ChaosConnector) Compensate(ctx context.Context, o connectors.Obligation) (connectors.CompensationResult, error) {
	// Compensation faults are not modelled in Phase 1 (matches Python/TS).
	return c.Inner.Compensate(ctx, o)
}

File: tape/sdk/typescript/src/chaos/connectors.ts
// ChaosConnector — declarative fault wrap around any registered connector.
// Mirrors `tape.chaos.connectors.ChaosConnector` (Python).

import type {
  Connector, EffectRecord, ObligationRecord,
  DispatchResult, ObservationResult, CompensationResult,
} from '../connectors/index.ts';
import type { Fault } from './scenarios.ts';

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

/**
 * Wraps an inner `Connector` and applies declarative faults around its
 * dispatch / observe methods. Three fault kinds are honoured:
 *
 *   lose_ack   — dispatch returns `unknown` after the inner call lands.
 *                Models a lost ack so the reconciler resolves via observe().
 *   duplicate  — observe returns `duplicate` instead of `confirmed`. Models
 *                the upstream having landed two rows for the same key.
 *   delay      — sleeps `ms ± jitter` before dispatch. Models slow upstreams.
 *
 * `rng` is the seeded PRNG from `Session` — same seed across two replays =
 * same fault sequence. The wrapper exposes the same `name` as the inner
 * connector so the outbox reactor routes to it transparently.
 */
export class ChaosConnector implements Connector {
  readonly name: string;
  readonly inner: Connector;
  readonly faults: readonly Fault[];
  private readonly rng: () => number;

  constructor(inner: Connector, faults: readonly Fault[], rng: () => number) {
    this.inner = inner;
    this.faults = faults;
    this.rng = rng;
    this.name = inner.name;
  }

  private fire(kind: string): Fault | null {
    for (const f of this.faults) {
      if (f.action !== kind) continue;
      if (f.probability >= 1.0 || this.rng() < f.probability) return f;
    }
    return null;
  }

  async dispatch(effect: EffectRecord): Promise<DispatchResult> {
    const d = this.fire('delay');
    if (d && d.ms > 0) {
      let ms = d.ms;
      if (d.jitter > 0) ms = Math.max(0, Math.floor(ms * (1.0 + (this.rng() * 2 - 1) * d.jitter)));
      await sleep(ms);
    }
    const result = await this.inner.dispatch(effect);
    if (result.outcome === 'confirmed' && this.fire('lose_ack')) {
      return {
        outcome: 'unknown',
        response: result.response,
        error: 'tape.chaos: simulated lost ack',
        dispatchId: result.dispatchId,
      };
    }
    return result;
  }

  async observe(effect: EffectRecord): Promise<ObservationResult> {
    const result = await this.inner.observe(effect);
    if (result.outcome === 'confirmed' && this.fire('duplicate')) {
      return {
        outcome: 'duplicate',
        response: result.response,
        count: (result.count ?? 1) + 1,
      };
    }
    return result;
  }

  async compensate(obligation: ObligationRecord): Promise<CompensationResult> {
    // Compensation faults are not modelled in Phase 1 (matches Python).
    return this.inner.compensate(obligation);
  }
}

/** Sugar: build a wrapped connector outside of a session. */
export function wrapConnector(inner: Connector, faults: readonly Fault[],
                                rng: () => number = Math.random): ChaosConnector {
  return new ChaosConnector(inner, faults, rng);
}
